src/design_notes/eager_drop.md

```diff
@@ -0,0 +1,67 @@
+# Eager drop design note
+
+- Project proposal [rust-lang/lang-team#86](https://github.com/rust-lang/lang-team/issues/86)
+
+## Observations
+
+### Any attempt to make drop run more eagerly will have to take borrows into account
+
+The original proposal was to use "CFG dead" as a criteria, but it's pretty clear that this will not work well. Example:
+
+```rust=
+{
+    let x = String::new();
+    let y = &x;
+    // last use of x is here
+    println!("{}", y);
+    // but we use y here
+    ...
+}
+```
+
+Here, the fact that `y` (indirectly) uses `x` feels like an important thing to take into account.
+
+### Some destructors can be run "at any time"...
+
+Some destructors have very significant side-effects. The most notable example is dropping a lock guard.
+
+Others correspond solely to "releasing resources": freeing memory is the most common example, but another might be replacing an entry in a table because you are done using it.
+
+### ...but sometimes that significance is only known at the call site
+
+However, it can be hard to know what is significant. For a lock guard, for example, if the lock is just being used to guard the data, then moving the lock release early is actually _desirable_, because you want to release the lock as soon as you are doing changing the data. But sometimes you have a `Mutex<()>`, in which case the lock has extra semantics. It's hard to know for sure.
+
+### Smarter drop placement will mean that adding uses of a variable changes when its destructor runs
+
+This is not necesarily a problem, but it's an obvious implication: right now, the drop always runs when we exit the scope, so adding further uses to a variable has no effect, but that would have to change. That could be surprising (e.g., adding a debug printout changes the time when a lock is released).
+
+In contrast, if you add an early drop `drop(foo)` today, you get helpful error messages when you try to use it again.
+
+In other words, it's useful to have the _destructor_ occurring at a known time (sometimes...).
+
+### Today's drop rules are, however, a source of confusion
+
+The fact that `let _ = foo` drops `foo` immediately is a known source of confusion, along with the lifetimes of temporaries in `match` statements and the like:
```

Comment:
This only drops the value of the expression `foo` when it is not a place expression. Using `foo` for the example looks too much like a place expression, so I find this example quite confusing (which, in a way, further strengthens the very point it wants to make 😀).

```suggestion
The fact that `let _ = foo()` drops the returned value immediately is a known source of confusion, along with the lifetimes of temporaries in `match` statements and the like:
```

I guess mentioning the following four different semantics for `let _{,g}` can further illustrate its being confusing.

  - `let _ = foo();`, which drops;
  - `let _g = foo();` which doesn't drop;
  - `let _ = foo;` which neither drops nor moves;
  - `let _g = foo;` which moves from `foo`.